### PR TITLE
Fix brackets issue on some extra edge cases

### PR DESCRIPTION
**Pull request by gigiblender**
#### Description

Describe the patch. What does it enhance? What does it fix?

This PR addresses the following issues; 
1) fix closing brackets on some edge cases
2) update end loop phi variables regardless the loop runs or not

#### Problem description

1) There are edge cases we did not cover during code gen and emitted the invalid code. 

2) If a loop never iterates and some phi variables of the loop are used after the loop, then we need to assign them outside the loop body. 

In the code below, `i_308` and `i_309` are never assigned a value when the loop does not iterate
```
i_363  =  i_195;
i_364  =  i_170;
for(;i_364 < i_111;) {
... loop body where we use `i_363` and `i_364` ....
i_308  =  i_363;
i_309  =  i_364;
}
... use `i_308` and `i_309` outside the loop...
```

To fix this, we emit the phi assignments after we close the loop scope.
```
i_363  =  i_195;
i_364  =  i_170;
for(;i_364 < i_111;) {
... loop body where we use `i_363` and `i_364` ....
}
i_308  =  i_363;
i_309  =  i_364;
... use `i_308` and `i_309` outside the loop...
```

#### Backend/s tested

- [X] OpenCL
- [ ] PTX

#### OS tested

- [X] Linux
- [ ] OSx
- [ ] Windows

#### Did you check on FPGAs?

If possible, check your changes on FPGAs.

- [ ] Yes
- [X] No

#### How to test the new patch?

All the unit tests pass, including slambench and the benchmarks.

### CODE REVIEW

**Comment by jjfumero** [comment]
Thanks @gigiblender , can you please check that this does not break the internal #731?

**Comment by gigiblender** [comment]
> Thanks @gigiblender , can you please check that this does not break the internal #731?

I checked the changes from that branch and ran all the tests (including the new FFT example). All looks good

File: drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/compiler/OCLBlockVisitor.java [hunk 1]
```diff
@@ -222,54 +226,85 @@ private void closeSwitchStatement(Block block) {
         }
     }
 
-    private boolean wasBlockAlreadyClosed(Block b) {
-        Block[] successors = b.getSuccessors();
-        for (Block s : successors) {
-            if (closedLoops.contains(s)) {
-                return true;
-            }
+    private boolean wasBlockAlreadyClosed(Block block) {
```
Comment (jjfumero) [inline drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/compiler/OCLBlockVisitor.java:229]: How is the map here? The counter does not mean that a particular block was already closed. Can you explain a bit?

File: drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/lir/OCLLIRStmt.java [hunk 2]
```diff
@@ -55,6 +55,37 @@ public final void emitCode(CompilationResultBuilder crb) {
 
     }
 
+    @Opcode("BEGIN_SKIP_SEQUENCE")
```
Comment (jjfumero) [inline drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/lir/OCLLIRStmt.java:58]: What are these "skip" instructions? Shouldn't be call NOOP instructions instead? 
In which cases we emit these instructions?

File: drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/compiler/OCLNodeLIRBuilder.java [hunk 3]
```diff
@@ -683,6 +683,26 @@ public void visitEndNode(final AbstractEndNode end) {
...
+            for (Block pred : curBlock.getPredecessors()) {
+                if (pred == loopBeginBlock) {
+                    shouldSkipStmt = true;
+                    break;
+                }
+            }
+        }
+
+        if (shouldSkipStmt) {
+            append(new OCLLIRStmt.BeginSkipStmt());
```
Comment (jjfumero) [inline drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/compiler/OCLNodeLIRBuilder.java:703]: So these instructions are just to keep control flow?

**Review by jjfumero — CHANGES_REQUESTED** [review]

**Comment by jjfumero** [comment]
All unitests + KFusiuonm-slambench are passing with this patch.

File: drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/compiler/OCLCompilationResultBuilder.java [hunk 4]
```diff
@@ -264,6 +265,33 @@ void emitLoopBlock(Block block) {
         emitBlock(block);
     }
 
+    void emitSkippedInstructions(Block block) {
```
Comment (jjfumero) [inline drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/compiler/OCLCompilationResultBuilder.java:268]: Document the need for "Skip" instructions. It looks to me more like a NOOP instruction.

**Review by jjfumero — CHANGES_REQUESTED** [review]

File: drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/lir/OCLLIRStmt.java [hunk 5]
```diff
@@ -55,6 +55,37 @@ public final void emitCode(CompilationResultBuilder crb) {
...
+    public static class BeginSkipStmt extends AbstractInstruction {
```
Comment (jjfumero) [inline drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/lir/OCLLIRStmt.java:59]: The semantics is to Relocate an instruction from within the loop to outside the loop. Therefore, let's update the name to `MarkRelocatebleInstruction`

File: drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/lir/OCLLIRStmt.java [hunk 6]
```diff
@@ -55,6 +55,37 @@ public final void emitCode(CompilationResultBuilder crb) {
...
+        }
+
+        @Override
+        public void emitCode(OCLCompilationResultBuilder crb, OCLAssembler asm) {
+            // No code is generated
+        }
+    }
+
+    @Opcode("END_SKIP_SEQUENCE")
+    public static class EndSkipStmt extends AbstractInstruction {
```
Comment (jjfumero) [inline drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/lir/OCLLIRStmt.java:74]: This LIR instruction can be deleted.

**Review by jjfumero — CHANGES_REQUESTED** [review]

File: drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/compiler/OCLBlockVisitor.java (hunk 1 above)
Comment (gigiblender) [reply drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/compiler/OCLBlockVisitor.java:229]: It is possible to have multiple nodes that can close a loop (blocks with `LoopExit` nodes). For such cases, we either generate an `else { break; } ` statement or we just close the loop body. To check if a loop has already been closed, I am using the `Map` above to compare against the number of loop exits.

File: drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/compiler/OCLNodeLIRBuilder.java (hunk 3 above)
Comment (gigiblender) [reply drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/graal/compiler/OCLNodeLIRBuilder.java]: Yes, if phi values are used before, inside, and after a loop, then we need to make sure that after the loop, the phi variables have been assigned (even if the loop did not iterate)

**Comment by gigiblender** [comment]
All changes addressed

**Review by jjfumero — APPROVED** [review]

**Review by stratika — APPROVED** [review]

**Pull request merged**